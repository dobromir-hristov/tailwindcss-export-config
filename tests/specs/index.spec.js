const ConvertTo = require('../../src/index')
const path = require('path')

const testConfig = require('../tailwind.config')
const testConfigDisabledPlugins = require('../tailwind-disabled-plugins.config')
const fse = require('fs-extra')

jest.mock('fs-extra')

describe('Tailwind Options Exporter', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('does not allow supplying unsupported formats', () => {
    expect(() => {
      new ConvertTo({
        config: testConfig,
        format: 'some_random',
        destination: 'doesnt_matter'
      })
    }).toThrowError(/not supported/)
  })

  it('converts the config by using the proper converter', () => {
    let converterInstance = new ConvertTo({
      config: testConfig,
      format: 'styl',
      destination: 'doesnt_matter'
    })

    expect(converterInstance.converterInstance.format).toBe('styl')

    converterInstance = new ConvertTo({
      config: testConfig,
      format: 'scss',
      destination: 'doesnt_matter'
    })

    expect(converterInstance.converterInstance.format).toBe('scss')
  })

  it('skips options that are disabled in `corePlugins` using the object pattern', () => {
    let converterInstance = new ConvertTo({
      config: testConfigDisabledPlugins,
      format: 'scss'
    })

    const scssConfig = converterInstance.convert()
    // assert it does not contain a few properties
    expect(scssConfig).not.toContain('borderRadius')
    expect(scssConfig).not.toContain('backgroundSize')
    // assert the whole snapshot
    expect(scssConfig).toMatchSnapshot()
  })

  it('skips options that are disabled in `corePlugins` using the array pattern', () => {
    let converterInstance = new ConvertTo({
      config: {
        ...testConfig,
        corePlugins: ['cursor']
      },
      format: 'scss'
    })

    const scssConfig = converterInstance.convert()
    // assert it does not contain a few properties
    expect(scssConfig).toContain('cursor')
    expect(scssConfig).not.toContain('backgroundSize')
    // assert the whole snapshot
    expect(scssConfig).toMatchSnapshot()
  })

  it('it properly includes the provided configuration properties', () => {
    let converterInstance = new ConvertTo({
      config: testConfig,
      format: 'scss',
      destination: 'doesnt_matter',
      flat: true
    })

    const scssConfig = converterInstance.convert()
    expect(scssConfig).toContain('$fontFamily-display: (Gilroy,sans-serif)')
    expect(scssConfig).toContain('$colors-cyan: #9cdbf')
  })

  it('allows using an object as a config', () => {
    let converterInstance = new ConvertTo({
      config: {},
      format: 'scss',
      destination: 'doesnt_matter'
    })

    expect(converterInstance.convert()).toMatchSnapshot('scss format')
  })

  it('allows using a path to an config', () => {
    let converterInstance = new ConvertTo({
      config: path.join(__dirname, '..', 'tailwind.config.js'),
      format: 'scss',
      destination: 'doesnt_matter'
    })
    expect(converterInstance.convert()).toMatchSnapshot('scss format')
  })

  it('writes the new config to a file', (done) => {
    fse.outputFile.mockImplementation(() => Promise.resolve())
    let converterInstance = new ConvertTo({
      config: path.join(__dirname, '../tailwind.config'),
      format: 'scss',
      destination: 'doesnt_matter'
    })
    converterInstance.writeToFile().then(() => {
      expect(fse.outputFile).toHaveBeenCalled()
      done()
    })
  })
})
