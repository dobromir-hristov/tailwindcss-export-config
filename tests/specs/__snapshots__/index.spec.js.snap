// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tailwind Options Exporter converts deeply nested configs, with flat:false: level-default 1`] = `
"/* Converted Tailwind Config to scss */
$customForms: (
  colors-blue: blue,
  colors-green: green,
  somethingElse-level1-color: pink,
  somethingElse-level1-arrayValue: (a,b,c),
  somethingElse-level1-nestedB-color: nested,
  somethingElse-level1-nestedB-padding: much,
  somethingElse-level1-nestedB-thing: thing,
  somethingElse-level1-nestedB-nestedC-color: nestedC,
  somethingElse-level1-nestedB-nestedC-nestedD-color: nestedD,
  somethingElse-level1-nestedB-nestedC-nestedD-color2: color2,
  somethingElse-level1-nestedB-nestedC-nestedD-nestedE-color: nestedE,
);

$customColors: (
  colors-blue: NotBlue,
  colors-green: notGreen,
);
"
`;

exports[`Tailwind Options Exporter converts deeply nested configs, with flat:false: level0 1`] = `
"/* Converted Tailwind Config to scss */
$customForms: (
  colors: (
    blue: blue,
    green: green,
  ),
  somethingElse: (
    level1-color: pink,
    level1-arrayValue: (a,b,c),
    level1-nestedB-color: nested,
    level1-nestedB-padding: much,
    level1-nestedB-thing: thing,
    level1-nestedB-nestedC-color: nestedC,
    level1-nestedB-nestedC-nestedD-color: nestedD,
    level1-nestedB-nestedC-nestedD-color2: color2,
    level1-nestedB-nestedC-nestedD-nestedE-color: nestedE,
  ),
);

$customColors: (
  colors: (
    blue: NotBlue,
    green: notGreen,
  ),
);
"
`;

exports[`Tailwind Options Exporter converts deeply nested configs, with flat:false: level2 1`] = `
"/* Converted Tailwind Config to scss */
$customForms: (
  colors: (
    blue: blue,
    green: green,
  ),
  somethingElse: (
    level1: (
      color: pink,
      arrayValue: (a,b,c),
      nestedB: (
        color: nested,
        padding: much,
        thing: thing,
        nestedC-color: nestedC,
        nestedC-nestedD-color: nestedD,
        nestedC-nestedD-color2: color2,
        nestedC-nestedD-nestedE-color: nestedE,
      ),
    ),
  ),
);

$customColors: (
  colors: (
    blue: NotBlue,
    green: notGreen,
  ),
);
"
`;

exports[`Tailwind Options Exporter converts deeply nested configs, with flat:false: level3 1`] = `
"/* Converted Tailwind Config to scss */
$customForms: (
  colors: (
    blue: blue,
    green: green,
  ),
  somethingElse: (
    level1: (
      color: pink,
      arrayValue: (a,b,c),
      nestedB: (
        color: nested,
        padding: much,
        thing: thing,
        nestedC: (
          color: nestedC,
          nestedD-color: nestedD,
          nestedD-color2: color2,
          nestedD-nestedE-color: nestedE,
        ),
      ),
    ),
  ),
);

$customColors: (
  colors: (
    blue: NotBlue,
    green: notGreen,
  ),
);
"
`;

exports[`Tailwind Options Exporter converts deeply nested configs, with flat:true 1`] = `
"/* Converted Tailwind Config to scss */
$customForms-colors-blue: blue;
$customForms-colors-green: green;
$customForms-somethingElse-level1-color: pink;
$customForms-somethingElse-level1-arrayValue: (a,b,c);
$customForms-somethingElse-level1-nestedB-color: nested;
$customForms-somethingElse-level1-nestedB-padding: much;
$customForms-somethingElse-level1-nestedB-thing: thing;
$customForms-somethingElse-level1-nestedB-nestedC-color: nestedC;
$customForms-somethingElse-level1-nestedB-nestedC-nestedD-color: nestedD;
$customForms-somethingElse-level1-nestedB-nestedC-nestedD-color2: color2;
$customForms-somethingElse-level1-nestedB-nestedC-nestedD-nestedE-color: nestedE;

$customColors-colors-blue: NotBlue;
$customColors-colors-green: notGreen;
"
`;

exports[`Tailwind Options Exporter skips options that are disabled in \`corePlugins\` using the array pattern 1`] = `
"/* Converted Tailwind Config to scss */
$cursor: (
  auto: auto,
  default: default,
  pointer: pointer,
  wait: wait,
  text: text,
  move: move,
  not-allowed: not-allowed,
);
"
`;
